rfcs/20190305-modular-tensorflow.md

```diff
@@ -0,0 +1,481 @@
+1.  TensorFlow Metapackage: This will be a thin package that only defines the composition of TensorFlow. Please see the detailed design section for more details on this package.
+1.  Optional TF packages: These packages will include the optional TF features users may choose to load and enable after they have TF working. Without these, TF will work just fine. Example features we will have as optional packages are GPU support, MKL support, or cloud filesystem support. These will use the C++ modules to load the functions they provide at runtime.
+
+
+## Detailed Design
+
+We will describe each key design element here in detail. To make the points clearer, trivial examples will be created.
+
+### Modularity in C/C++
+
+This section will describe the key design points for C++ modularity.
+
+#### Modules
+
+Each module's main pieces will be a module interface in C, and a registry for plugins implementing this module. As a supporting piece, each module will also need to provide a mechanism for plugins to add themselves to the registry at runtime. Below is a toy example for the described design:
+
+
+```
+// A simple toy module called "M"
+typedef struct M_context;
+
+// The summary of the interface plugins need to implement
+typedef struct MInterface {
+  void (*M_Init)(M_context*, string);
+  void (*M_Execute)(M_context*);
+};
+
+// The plugin registry for module M
+// The registry here implemented uses C++ class std::map
+// This is OK as it is not exposed in our example here.
+// As far as module implementations are concerned, they only
+// need to see and use M_RegisterPlugin method, and that method takes // care of their addition into this registry.
+std::map<string, MInterface> m_plugin_registry;
+
+// Function to call for each plugin at load time
+// to add themselves to the registry
+void M_RegisterPlugin(
+    string plugin, 
+    void (*M_Init)(M_context*, string),
+    void (*M_Execute)(M_context*)) {
+  // Omitting error handling.
+  m_plugin_registry.insert(plugin,
+                           MInterface(M_Init, M_Execute));
+}
+
+// Implementation of the interface is just a thin layer to
+// get the correct plugin and call it.
+// Here we assume that plugin explicitly gets selected by
+// the init function. Some modules can go with implicit selection,
+// Such as deducing the filesystem from the file path.
+void M_Init(M_context* ctx, string id) {
+  // Just a quick hacky way to select the plugin here
+  ctx->interface = m_plugin_registry[id];
+  ctx->interface.M_Init(ctx, id);
+}
+
+void M_Execute(M_context* ctx) {
+  ctx->interface.M_Execute(ctx);
+}
+```
+
+
+**Please note that the above is a toy example to explain the abstract ideas described above. Exact implementation can vary across different modules.**
+
+Interface has to be pure C at the ABI level. We can have C++ header-only libraries built on top of these C ABI/APIs.
```

Comment:
Great idea to offer header-only C++ libraries that will make C ABI more accessible.

Reply:
So is the official c/c++ API just a core API?
I.e. If I model/train in python with the new "use Keras in TF 2.0" mantra/marketing I am not sure that I can use the same API in c++ to integrate inference in a c++ app.
I still not find a way on how we can achieve this in the roadmap.
Train with python API, quickly export the model and load it in c++, make some ops call in c++ at inference. 
I.e. Just to check the competitor I see Aten has a very good roadmap to stay on par with python API https://pytorch.org/cppdocs/ and Torchscript makes trivial to import a just trained model in c++ on the fly with just with a one or two Trocscript call.